pkg/pgclient/config.go

```diff
@@ -154,35 +159,81 @@ func (cfg *Config) GetConnectionStr() string {
+	cfgMaxConns := cfg.cfgMaxConns(dbMaxConns)
+
+	// calc maint pool size
+	maintPoolSize = cfg.MaintPoolSize
+	if maintPoolSize == defaultPoolSize {
+		maintPoolSize = defaultMaintPoolSize
+	}
+
+	// calc writer pool size
+	if !readOnly {
+		writerPoolSize = cfg.calcPoolSize(dbMaxConns, cfg.WriterPoolSize, defaultConnFraction)
+	}
+
+	// calc reader pool size
+	readerFraction := defaultReaderFraction
+	if readOnly {
+		readerFraction = defaultConnFraction
+	}
+	readerPoolSize = cfg.calcPoolSize(dbMaxConns, cfg.ReaderPoolSize, readerFraction)
+	if cfg.ReaderPoolSize == defaultPoolSize && readerPoolSize+writerPoolSize+maintPoolSize > cfgMaxConns {
+		readerPoolSize = readerPoolSize - maintPoolSize
+	}
+
+	err = cfg.validatePoolSizes(readOnly, dbMaxConns, cfgMaxConns, readerPoolSize, writerPoolSize, maintPoolSize)
 	if err != nil {
-		return 0, fmt.Errorf("max connections: %w", err)
+		return 0, 0, 0, err
 	}
+	return readerPoolSize, writerPoolSize, maintPoolSize, nil
+}
 
+func (cfg *Config) cfgMaxConns(dbMaxConns int) int {
+	// if db.connections-max is the default, use 80% of the postgres max_connections setting
+	cfgMaxConns := cfg.MaxConnections
+	if cfgMaxConns == defaultMaxConns {
+		cfgMaxConns = int(float64(dbMaxConns) * defaultMaxConnsPercentage)
+	}
+	if cfgMaxConns > dbMaxConns {
+		log.Warn("msg", "configured db.connections-max is greater then postgres max_connections",
+			"db.connections-max", cfgMaxConns,
+			"max_connections", dbMaxConns,
+		)
+	}
+	return cfgMaxConns
+}
+
+func (cfg *Config) validatePoolSizes(readOnly bool, dbMaxConns, cfgMaxConns, readerPoolSize, writerPoolSize, maintPoolSize int) error {
 	switch {
-	case inputPoolSize == defaultPoolSize:
-		// For the default case, we need to take up defaultFraction of allowed connections.
-		poolSize := float64(maxConns) * defaultFraction
+	case maintPoolSize < MinPoolSize:
+		return fmt.Errorf("maint pool size canot be less than %d: received %d", MinPoolSize, maintPoolSize)
+	case !readOnly && writerPoolSize < MinPoolSize:
+		return fmt.Errorf("writer pool size canot be less than %d: received %d", MinPoolSize, writerPoolSize)
+	case readerPoolSize < MinPoolSize:
+		return fmt.Errorf("reader pool size canot be less than %d: received %d", MinPoolSize, readerPoolSize)
+	case readerPoolSize+writerPoolSize+maintPoolSize > dbMaxConns:
+		return fmt.Errorf("reader-pool (size=%d) + writer-pool (size=%d) + maint-pool (size=%d) more than postgres max_connections (%d). Decrease the pool-sizes", readerPoolSize, writerPoolSize, maintPoolSize, dbMaxConns)
+	case readerPoolSize+writerPoolSize+maintPoolSize > cfgMaxConns:
+		return fmt.Errorf("reader-pool (size=%d) + writer-pool (size=%d) + maint-pool (size=%d) more than db.connections-max (%d). Increase the db.connections-max or decrease the pool-sizes", readerPoolSize, writerPoolSize, maintPoolSize, cfgMaxConns)
+	}
+	return nil
```

Comment:
Minor nit: this is a minor semantic thing but putting the return nil into default case lets the reader know there is nothing after the switch. This forces you to do returns in every case of the switch statement (which is what we want here).

Reply:
👍